File: ECnotes/Sem2/LivExamples/VB/Ch19CartProfile/App_Code/ShoppingCart.vb
Imports Microsoft.VisualBasic

<Serializable()> _
Public Class ShoppingCart

    Private Cart As SortedList = New SortedList()

    Public ReadOnly Property Count() As Integer
        Get
            Return Cart.Count
        End Get
    End Property

    Public Sub AddItem(ByVal Item As CartItem)
        Dim sProductID As String = Item.Product.ProductID
        If (Cart.ContainsKey(sProductID)) Then
            Dim ExistingItem As CartItem
            ExistingItem = CType(Cart.Item(sProductID), CartItem)
            ExistingItem.Quantity += Item.Quantity
        Else
            Cart.Add(sProductID, Item)
        End If
    End Sub

    Public Function GetItems() As CartItem()
        Dim Items(Cart.Count) As CartItem
        Dim i As Integer
        If (Cart.Count > 0) Then
            For i = 0 To Cart.Count - 1
                Items(i) = CType(Cart.GetByIndex(i), CartItem)
            Next
        End If
        Return Items
    End Function

    Public Sub Clear()
        Cart.Clear()
    End Sub

    Public Sub RemoveAt(ByVal i As Integer)
        Cart.RemoveAt(i)
    End Sub

End Class

File: ECnotes/Sem2/LivExamples/CS/Ch19CartProfile/App_Code/ShoppingCart.cs
using System;
using System.Collections;

// You must mark this class as serializable
// to get the Profile feature to work
[Serializable]
public class ShoppingCart
{
	SortedList cart = new SortedList();

    public int Count
    {
        get
        {
            return cart.Count;
        }
	}

	public void AddItem(CartItem item)
	{
		string productID = item.Product.ProductID;

		if (cart.ContainsKey(productID))
		{
			CartItem existingItem = (CartItem) cart[productID];

			existingItem.Quantity += item.Quantity;
		}
		else
			cart.Add(productID, item);
	}

	public CartItem[] GetItems()
	{
		CartItem[] items = new CartItem[cart.Count];

		for (int i = 0; i < cart.Count; i++)
			items[i] = (CartItem) cart.GetByIndex(i);

		return items;
	}

	public void Clear()
	{
		cart.Clear();
	}

	public void RemoveAt(int i)
	{
		cart.RemoveAt(i);
	}
}